whisper/src/net.rs

```diff
@@ -0,0 +1,638 @@
+
+		let messages = self.messages.read();
+		let peers = self.peers.read();
+
+		// send each peer a packet with new messages it may find relevant.
+		for (peer_id, peer) in peers.iter() {
+			let mut peer_data = peer.lock();
+			peer_data.note_evicted(&pruned_hashes);
+
+			let punish_timeout = |last_activity: &SystemTime| {
+				if *last_activity + Duration::from_millis(MAX_TOLERATED_DELAY_MS) <= now {
+					debug!(target: "whisper", "Disconnecting peer {} due to excessive timeout.", peer_id);
+					io.disconnect_peer(*peer_id);
+				}
+			};
+
+			// check timeouts and skip peers who we can't send a rally to.
+			match peer_data.state {
+				State::Unconfirmed(ref time) | State::TheirTurn(ref time) => {
+					punish_timeout(time);
+					continue;
+				}
+				State::OurTurn => {}
+			}
+
+			// construct packet, skipping messages the peer won't accept.
+			let mut stream = RlpStream::new();
+			stream.begin_unbounded_list();
+
+			for message in messages.iter() {
+				if !peer_data.will_accept(message) { continue }
+
+				if stream.estimate_size(message.encoded_size()) > MAX_MESSAGES_PACKET_SIZE {
+					break;
+				}
+
+				peer_data.note_known(message);
+				stream.append(message.envelope());
+			}
+
+			stream.complete_unbounded_list();
+
+			peer_data.state = State::TheirTurn(SystemTime::now());
+			if let Err(e) = io.send(*peer_id, packet::MESSAGES, stream.out()) {
+				debug!(target: "whisper", "Failed to send messages packet to peer {}: {}", peer_id, e);
+				io.disconnect_peer(*peer_id);
+			}
+		}
+	}
+
+	// handle status packet from peer.
+	fn on_status(&self, peer: &PeerId, status: UntrustedRlp)
+		-> Result<(), Error>
+	{
+		let proto: usize = status.as_val()?;
+		if proto != PROTOCOL_VERSION { return Err(Error::ProtocolVersionMismatch(proto)) }
+
+		let peers = self.peers.read();
+		match peers.get(peer) {
+			Some(peer) => {
+				let mut peer = peer.lock();
+				let our_node_key = self.node_key.read().clone();
+
+				// handle this basically impossible edge case gracefully.
+				if peer.node_key == our_node_key {
+					return Err(Error::SameNodeKey);
+				}
+
+				// peer with lower node key begins the rally.
+				if peer.node_key > our_node_key {
+					peer.state = State::OurTurn;
+				} else {
+					peer.state = State::TheirTurn(SystemTime::now());
+				}
+
+				Ok(())
+			}
+			None => {
+				debug!(target: "whisper", "Received message from unknown peer.");
+				Err(Error::UnknownPeer(*peer))
+			}
+		}
+	}
+
+	fn on_messages(&self, peer: &PeerId, message_packet: UntrustedRlp)
+		-> Result<(), Error>
+	{
+		let mut messages_vec = {
+			let peers = self.peers.read();
+			let peer = match peers.get(peer) {
+				Some(peer) => peer,
+				None => {
+					debug!(target: "whisper", "Received message from unknown peer.");
+					return Err(Error::UnknownPeer(*peer));
+				}
+			};
+
+			let mut peer = peer.lock();
+
+			if !peer.can_send_messages() {
+				return Err(Error::UnexpectedMessage);
```

Comment:
Can connection recover after this? Maybe it would be better to disconnect on any such error?